lib/ansible/modules/cloud/amazon/kms.py

```diff
@@ -0,0 +1,272 @@
+                changes_needed[granttype] = 'remove'
+                if not dry_run:
+                    statement['Principal']['AWS'].remove(role_arn)
+
+    if len(changes_needed) and not dry_run:
+        #policy_json_string = json.dumps({'Policy': policy})
+        policy_json_string = json.dumps(policy)
+        kms.put_key_policy(KeyId=keyarn, PolicyName='default', Policy=policy_json_string)
+
+        # returns nothing, so we have to just assume it didn't throw
+        ret['changed'] = True
+
+    ret['changes_needed'] = changes_needed
+    if dry_run:
+        # true if changes > 0
+        ret['changed'] = (not len(changes_needed) == 0)
+
+    return ret
+
+def assert_policy_shape(policy):
+    '''Since the policy seems a little, uh, fragile, make sure we know approximately what we're looking at.'''
+    errors = []
+    if policy['Version'] != "2012-10-17":
+        errors.append('Unknown version/date ({}) of policy. Things are probably different than we assumed they were.'.format(policy['Version']))
+
+    haz = {}
+    for statement in policy['Statement']:
+        for label,sidlabel in statement_label.items():
+            if statement['Sid'] == sidlabel:
+                haz[label] = True
+
+    for statementtype in statement_label.keys():
+        if not haz.get(statementtype):
+            errors.append('Policy is missing {}.'.format(statementtype))
+
+    if len(errors):
+        raise Exception('Problems asserting policy shape. Cowardly refusing to modify it: {}'.format(' '.join(errors)))
+    return None
+
+def main():
+    argument_spec = ansible.module_utils.ec2.ec2_argument_spec()
+    argument_spec.update(dict(
+            mode = dict(choices=['grant', 'deny'], default='grant'),
+            key_alias = dict(required=False, type='str'),
+            key_arn = dict(required=False, type='str'),
+            role_name = dict(required=False, type='str'),
+            role_arn = dict(required=False, type='str'),
+            grant_types = dict(required=False, type='list'),
+        )
+    )
+
+    module = AnsibleModule(
+        supports_check_mode=True,
+        argument_spec=argument_spec,
+        required_one_of=[['key_alias', 'key_arn'], ['role_name', 'role_arn']],
+        required_if=[['mode', 'grant', ['grant_types']]]
+    )
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 required for this module')
+
+    result = {}
+    mode = module.params['mode']
+
+
+    try:
+        region, ec2_url, aws_connect_kwargs = ansible.module_utils.ec2.get_aws_connection_info(module, boto3=True)
+        kms = ansible.module_utils.ec2.boto3_conn(module, conn_type='client', resource='kms', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        iam = ansible.module_utils.ec2.boto3_conn(module, conn_type='client', resource='iam', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoCredentialsError as e:
+        module.fail_json(msg=str(e))
```

Comment:
okay. Looks like usage of this for the "create boto conn" exception is all over the map, fwiw.,